pkg/didcomm/dispatcher/inbound/inbound_message_handler.go

```diff
@@ -0,0 +1,238 @@
+	kaIdentifier = "#"
+)
+
+// MessageHandler handles inbound envelopes, processing then dispatching to a protocol service based on the
+// message type.
+type MessageHandler struct {
+	didConnectionStore     did.ConnectionStore
+	msgSvcProvider         api.MessageServiceProvider
+	services               []dispatcher.ProtocolService
+	getDIDsBackOffDuration time.Duration
+	getDIDsMaxRetries      uint64
+	messenger              service.InboundMessenger
+	initialized            bool
+}
+
+type provider interface {
+	DIDConnectionStore() did.ConnectionStore
+	MessageServiceProvider() api.MessageServiceProvider
+	AllServices() []dispatcher.ProtocolService
+	GetDIDsBackOffDuration() time.Duration
+	GetDIDsMaxRetries() uint64
+	InboundMessenger() service.InboundMessenger
+}
+
+// NewInboundMessageHandler creates an inbound message handler, that processes inbound message Envelopes,
+// and dispatches them to the appropriate ProtocolService.
+func NewInboundMessageHandler(p provider) *MessageHandler {
+	h := MessageHandler{}
+	h.Initialize(p)
+
+	return &h
+}
+
+// Initialize initializes the MessageHandler. Any call beyond the first is a no-op.
+func (handler *MessageHandler) Initialize(p provider) {
+	if handler.initialized {
+		return
+	}
+
+	handler.didConnectionStore = p.DIDConnectionStore()
+	handler.msgSvcProvider = p.MessageServiceProvider()
+	handler.services = p.AllServices()
+	handler.getDIDsBackOffDuration = p.GetDIDsBackOffDuration()
+	handler.getDIDsMaxRetries = p.GetDIDsMaxRetries()
+	handler.messenger = p.InboundMessenger()
+
+	handler.initialized = true
+}
+
+// HandleInboundEnvelope handles an inbound envelope, dispatching it to the appropriate ProtocolService.
+func (handler *MessageHandler) HandleInboundEnvelope(envelope *transport.Envelope) error { // nolint:gocyclo,funlen
+	var (
+		msg service.DIDCommMsgMap
+		err error
+	)
+
+	msg, err = service.ParseDIDCommMsgMap(envelope.Message)
+	if err != nil {
+		return err
+	}
+
+	isV2, err := service.IsDIDCommV2(&msg)
+	if err != nil {
+		return err
+	}
+
+	// if msg is a didcomm v2 message, do additional handling
+	if isV2 {
+		logger.Debugf("didcomm v2 message handling before service...")
+	}
+
+	// find the service which accepts the message type
+	for _, svc := range handler.services {
+		if svc.Accept(msg.Type()) {
+			var myDID, theirDID string
+
+			switch svc.Name() {
+			// perf: DID exchange doesn't require myDID and theirDID
+			case didexchange.DIDExchange:
+			default:
+				myDID, theirDID, err = handler.getDIDs(envelope)
+				if err != nil {
+					return fmt.Errorf("inbound message handler: %w", err)
+				}
+			}
+
+			_, err = svc.HandleInbound(msg, service.NewDIDCommContext(myDID, theirDID, nil))
+
+			return err
+		}
+	}
+
+	// in case of no services are registered for given message type,
+	// find generic inbound services registered for given message header
+	for _, svc := range handler.msgSvcProvider.Services() {
+		h := struct {
+			Purpose []string `json:"~purpose"`
```

Comment:
is this a V1 format?

Reply:
Yes - I'm not sure what will happen with the message service and purpose decorator when/if it gets ported to didcomm/v2, but this part is v1 only.

I can put it inside an `if`.

Reply:
yes, it would be better

Reply:
done